hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,171 @@
+package com.hazelcast.spi.impl;
+
+import com.hazelcast.instance.Node;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.util.ConcurrencyUtil;
+import com.hazelcast.util.ConstructorFunction;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Monitors the operation threads to find and collect data of long running operations.
+ */
+final class SlowOperationDetector extends Thread {
+
+    private static final long ONE_SECOND_IN_NANOS = TimeUnit.SECONDS.toNanos(1);
+
+    private final BasicOperationScheduler.OperationThread[] genericOperationThreads;
+    private final BasicOperationScheduler.OperationThread[] partitionOperationThreads;
+
+    private final CurrentOperationData[] genericCurrentOperationData;
+    private final CurrentOperationData[] partitionCurrentOperationData;
+
+    private final long slowOperationThresholdNanos;
+    private final long logPurgeIntervalNanos;
+    private final long logLifetimeNanos;
+
+    private final ILogger logger;
+
+    private final ConcurrentHashMap<Integer, SlowOperationLog> slowOperationLogs = new ConcurrentHashMap<Integer, SlowOperationLog>();
+
+    private long nowNanos;
+    private long nowMillis;
+
+    private volatile boolean running = true;
+
+    public SlowOperationDetector(Node node, BasicOperationScheduler scheduler) {
+        super(SlowOperationDetector.class.getSimpleName());
+        setDaemon(true);
+
+        genericOperationThreads = scheduler.genericOperationThreads;
+        partitionOperationThreads = scheduler.partitionOperationThreads;
+
+        genericCurrentOperationData = new CurrentOperationData[genericOperationThreads.length];
+        partitionCurrentOperationData = new CurrentOperationData[partitionOperationThreads.length];
+
+        initCurrentOperationData(genericCurrentOperationData);
+        initCurrentOperationData(partitionCurrentOperationData);
+
+        int thresholdMillis = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS.getInteger());
+        int purgeIntervalSeconds = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_LOG_PURGE_INTERVAL_SECONDS.getInteger());
+        int logLifetimeSeconds = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_LOG_RETENTION_SECONDS.getInteger());
+
+        slowOperationThresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMillis);
+        logPurgeIntervalNanos = TimeUnit.SECONDS.toNanos(purgeIntervalSeconds);
+        logLifetimeNanos = TimeUnit.SECONDS.toNanos(logLifetimeSeconds);
+
+        logger = node.getLogger(BasicOperationScheduler.class);
+    }
+
+    @Override
+    public void run() {
+        long lastLogPurge = System.nanoTime();
+
+        while (running) {
+            nowNanos = System.nanoTime();
+            nowMillis = System.currentTimeMillis();
```

Comment:
The `System.nanoTime()` is to measure the duration of an operation, because it's the most accurate method we have. Since it has nothing to do with our real time `System.currentTimeMillis()` is used to calculate the `startedAt` of an operation (so we have a timestamp which a user can read in the management center).

It doesn't matter if the values are a bit off. But they are meant as now (especially for the `startedAt` calculation I think it is much easier to read than a start/begin without having an end).